network/p2p/dns/cache.go

```diff
@@ -0,0 +1,88 @@
+package dns
+
+import (
+	"net"
+	"sync"
+	"time"
+)
+
+// defaultTimeToLive is the default duration a dns result is cached.
+const defaultTimeToLive = 5 * time.Minute
```

Comment:
Another thing to consider is to make the DNS cache entry TTL active instead of passive:

With passive expiry the TTL times out but the timeout is not discovered until lookup time, and then this causes an annoying extra delay for the associated connection using the domain.

Passive expiry makes sense if you don't know that the domain will ever be used again.

With active expiry the TTL times out and the timeout is noticed immediately and looked up again immediately even though there is no associated connection using the domain or waiting for the domain to be resolved.

Active expiry makes sense if you know it's highly likely that the domain will be used again.

With active expiry then typically the TTL used to expire the cached DNS record is a few seconds before the real TTL. In this way, the DNS record should always be in the cache and never expired when it's time to use it :-)

Reply:
Thanks for the suggestion, while I was developing this DNS cache, I faced a similar design decision and went over this tradeoff between the granularity and lock contention. 

- **Finer granularity and higher lock contention:** since we have a separate `timestamp` per cached entry, there is no _global_ expiry time for our cache entries, and each entry is _expired_ independently. Hence, we need to iteratively go over the cache using a timer (e.g., every 1 min), checking them all, and invalidate expired ones. This would add a constant lock contention that I was originally trying to avoid since every once in a while, the entire cache is locked for examining its entries against expiry. 
- **Coarser granularity and lower lock contention:** we can go simpler assigning a global timeout for the entire cache, i.e., regardless of when an entry has been cached, we invalidate the entire cache once in a while, e.g., every one minute. This would make the lock contention less as invalidating the entire cache goes faster than checking each entry separately against expiry. Of course we sacrifice the granularity of `timestamp`ing in return for lower lock contention. This solution works well assuming that DNS lookups are following a uniform distribution with time. However, in practice, the application layer goes independently of the DNS cache timer. Hence, it is likely to have lots of DNS lookup requests very close to the global timer expiry. In that case, all those (very recently) cached lookups are expired pre-maturely when the timer expires, e.g., on a 1-minute timer, all DNS lookups results cached in 50th second are also invalidated despite they are very fresh. This poses a concern if there is a drift in the application layer that the majority of DNS lookups happen close the timer expiry. 

Since this tradeoff needed some prior assumptions about the distribution of DNS lookup queries, I went with the simplest solution of passive expiry. Please let me know if you have a strong recommendation towards an active expiry (or any other solution).

Reply:
@yhassanzadeh13 Thanks for replying to the comment! :-)

Regarding the 'active' detection of timeouts:

One way I have seen this implemented in other systems is to keep the map values in chronological order. This means the cache entries not only exist in the map, but there is also a double linked list between them so that they are in chronological order. So the ends of the double linked list represent the first and last to 'timeout'. This is easier to implement if all the timeouts have the same TTL, and has two advantages:

1. When the timer (e.g. every 1 min) awakens then you only have to check the earliest TTL and if that does not timeout then you know none of the others timeout out as well.

2. If, upon the timer, the earliest TTL has timed out then it is not necessary to iterate over all cache entries to check their TTL, instead you can just go to the known next earliest TTL and check that.

Also, using the RW lock it's possible to iterate over the timeout 'tail' using the concurrent 'read' lock. The blocking 'write' lock is only necessary when it comes to actually updating the data structure :-)

This is one way I've seen that systems work efficiently which have very many similar timeouts, e.g. a socket server with 1M TCP connections each with a timeout. Instead of having 1M timers, or iterating over 1M TCP connection data structures to detect timeouts, the data structures are kept in chronological order of timeout.

Regarding the expected cache hit / miss ratio and how it effects latency:

If there are `n` unique domains (guessing this is ~ 300 on `mainnet`?) used frequency `f` times (guessing this is 1?) every `s` seconds (guessing this is 10 minutes due to connection purging) and they are cached for `ttl` seconds (according to this review the hard-coded TTL is 5 minutes; i.e. half of the 10 minute connection purge) and the non-cached lookup time is `ms` ms (let's say this is 10 ms for arguments sake), then:

* If passive timeouts: If `ttl` < `s` then we'll save ((`n` * `f`) - `n`) * `ms` ms at connection time, or? Which is zero ms because all cached domains will timeout out before the next connection purge? <-- Note: This is why I'm worried the current PR will not help reduce DNS requests.
* If active timeouts: If `ttl` < `s` then we'll save `n` * `ms` ms at connection time, or? Because the re-caching will be done independently of the connection during active timeout?

Let's say we are using passive timeouts, and the `ttl` is 15 minutes which is longer than `s` which is 10 minutes. In this case, every even 10 minutes all domains would be cached, but every odd 10 minutes all domains would timeout. So we'd only be using cached domains 50% of the time, or?

So with DNS caching and passive timeouts then the connection time savings only occur if the `ttl` is a multiple of the usage frequency `s`, and to make the savings meaningful, this needs to be larger, e.g. if `s` is 10 minutes, then `ttl` is ~ hours? But there would still be 10 minute purges where ALL the domains are effectively looked up?

> prior assumptions about the distribution of DNS lookup queries

Would it make sense to collect data about the distribution of the DNS lookups over say, an hour, or ~ 6 connection purges, in order to predict / simulate the expected cache hit / miss ratio? And then choose TTLs & active / passive timeouts based upon those real world observations?